### ISSUE TITLE
Segfault in SetDepedency

### ISSUE THREAD

**Issue by Kh4L** [apache/incubator-mxnet#16806]
## Description
GluonCV Mask-RCNN training script segfaults in SetDepency when using `mrcnn_mask_target` op in symbolic mode (with the hybridized model). 

`master`s  `mrcnn_mask_target`  op do not work but it is fixed it here https://github.com/apache/incubator-mxnet/pull/16673/files 

### Error Message
```
Segfault
#0  0x00007ffe405bcefc in mxnet::imperative::SetDependency(nnvm::NodeAttrs const&, mxnet::Context const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> >*, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> >*, std::vector<mxnet::Resource, std::allocator<mxnet::Resource> >*, std::vector<unsigned int, std::allocator<unsigned int> >*, mxnet::DispatchMode) () from /usr/local/lib/libmxnet.so
#1  0x00007ffe405a7e15 in mxnet::Imperative::InvokeOp(mxnet::Context const&, nnvm::NodeAttrs const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&, mxnet::DispatchMode, mxnet::OpStatePtr) () from /usr/local/lib/libmxnet.so
#2  0x00007ffe4059af18 in std::_Function_handler<void (mxnet::OpStatePtr const&), mxnet::imperative::RunGraph(bool, nnvm::IndexedGraph const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, unsigned long, unsigned long, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> >&&, std::vector<unsigned int, std::allocator<unsigned int> >&&, std::vector<mxnet::OpStatePtr, std::allocator<mxnet::OpState
Ptr> >*, std::vector<mxnet::DispatchMode, std::allocator<mxnet::DispatchMode> > const&, bool, std::vector<mxnet::TShape, std::allocator<mxnet::TShape> >*)::{lambda(mxnet::OpStatePtr const&)#1}>::_M_invoke(std::_Any_data const&, mxnet::OpStatePtr const&) () from /usr/local/lib/libmxnet.so
#3  0x00007ffe4059b298 in (anonymous namespace)::InvokeOperator(nnvm::IndexedGraph const&, int, bool, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, mxnet::Context, std::vector<mxnet::OpStatePtr, std::allocator<mxnet::OpStatePtr> >*, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> >*, std::vector<unsigned int, std::allocator<unsigned int> >*, std::function<void (mxnet::OpStatePtr const&)>) () from /usr/local/lib/libmxnet.so
#4  0x00007ffe4059cb15 in mxnet::imperative::RunGraph(bool, nnvm::IndexedGraph const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, unsigned long, unsigned long, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> >&&, std::vector<unsigned int, std::allocator<unsigned int> >&&, std::vector<mxnet::OpStatePtr, std::allocator<mxnet::OpStatePtr> >*, std::vector<mxnet::DispatchMode, std::allocator<mxnet::DispatchMode> > const&, bool, std::vector<mxnet::TShape, std::allocator<mxnet::TShape> >*) () from /usr/local/lib/libmxnet.so
#5  0x00007ffe405b3571 in mxnet::Imperative::Backward(std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, bool, bool, bool) () from /usr/local/lib/libmxnet.so
```

### Steps to reproduce

1. Compile MXNet on this branch https://github.com/Kh4L/incubator-mxnet/tree/enable_sym_mask_symbol
2. Get GluonCV on `master`
4. pip install --upgrade gluon-cv/
3. Run:
python ./gluon-cv/scripts/instance/mask_rcnn/train_mask_rcnn.py --gpus 0 --dataset coco -j 4 --batch-size 1

## What have you tried to solve it?

1. Adding `symbol.stop_gradient`: does not help.
2. Debugging with GDB and prints: 

The segfault does not occur during the `mask_target_op` execution but it happens later in `SetDepedency` of a `zeros_like` op.
More precisely on this line: https://github.com/apache/incubator-mxnet/blob/a37dcd46cba88dfd0478b601fb8fbf9b425780c0/src/imperative/imperative_utils.h#L288

I tried dereferencing `i` et and calling the `NDArray::var()` separately and they both look fine. 
So it looks like it happens when `i` is being pushed to `read_vars`... the segfault seems to be happening in the copy ctor of the `engine::VarHandle`?

**Comment by zhreshold** [comment]
Can you minimize the reproduction code? For example, would a simple snippet like this reproduce segfault?

```python
with autograd.record():
  mask_targets, mask_masks = mrcnn_mask_target(rois, gt_masks, matches, cls_targets)
```

**Comment by samskalicky** [comment]
@lanking520 assign @zhreshold
